File: tests/cross-verification/nasam/_gen/gen.py
#!/usr/bin/env python3
# Independent Python hand-port oracle for Pelle Evensen's `nasam` mixer.
# Re-implements the public-domain reference FROM SCRATCH — no shared module, no IR —
# so it is a genuine N-way peer to the generated-from-ir TS oracle.
# Reference: https://mostlymangling.blogspot.com/2020/01/nasam-not-another-strange-acronym-mixer.html
#   x ^= ror(x,25) ^ ror(x,47);  x *= 0x9E6C63D0676A9A99;  x ^= x>>23 ^ x>>51;
#   x *= 0x9E6D62D06F6A9A9B;      x ^= x>>23 ^ x>>51
import json
import os

MASK = (1 << 64) - 1
M1 = 0x9E6C63D0676A9A99
M2 = 0x9E6D62D06F6A9A9B


def ror(x: int, r: int) -> int:
    return ((x >> r) | (x << (64 - r))) & MASK


def nasam(x: int) -> int:
    x &= MASK
    x ^= ror(x, 25) ^ ror(x, 47)
    x = (x * M1) & MASK
    x ^= (x >> 23) ^ (x >> 51)
    x = (x * M2) & MASK
    x ^= (x >> 23) ^ (x >> 51)
    return x & MASK


INPUTS = {
    "x-0": 0,
    "x-1": 1,
    "x-2": 2,
    "x-10": 10,
    "x-255": 255,
    "x-u64max": (1 << 64) - 1,
    "x-golden": 11400714819323198485,
    "x-2pow63": 1 << 63,
    "x-12345678901234567890": 12345678901234567890,
    "x-1e18": 10 ** 18,
}

out = {"_source": "hand-port-python"}
for k, v in INPUTS.items():
    out[k] = str(nasam(v))

target = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "python-output.json")
with open(target, "w") as f:
    f.write(json.dumps(out, indent=2) + "\n")
print("wrote python-output.json (hand-port)")

File: tests/cross-verification/xoshiro256ss/_gen/gen.rs
// Independent Rust hand-port oracle for the xoshiro256** OUTPUT SCRAMBLER.
// Re-implements result = rotl(x*5, 7) * 9 (width 64) FROM SCRATCH — a genuine
// N-way peer. wrapping_mul / rotate_left give native mod-2^64 wrap.
// Public-domain reference: https://prng.di.unimi.it/xoshiro256starstar.c
use std::fs;
use std::path::Path;

fn scramble(x: u64) -> u64 {
    x.wrapping_mul(5).rotate_left(7).wrapping_mul(9)
}

fn main() {
    let inputs: [(&str, u64); 10] = [
        ("x-0", 0),
        ("x-1", 1),
        ("x-2", 2),
        ("x-10", 10),
        ("x-255", 255),
        ("x-u64max", 18446744073709551615),
        ("x-golden", 11400714819323198485),
        ("x-2pow63", 9223372036854775808),
        ("x-12345678901234567890", 12345678901234567890),
        ("x-1e18", 1000000000000000000),
    ];

    let mut s = String::from("{\n");
    s.push_str("  \"_source\": \"hand-port-rust\",\n");
    for (i, (id, x)) in inputs.iter().enumerate() {
        let comma = if i == inputs.len() - 1 { "" } else { "," };
        s.push_str(&format!("  \"{}\": \"{}\"{}\n", id, scramble(*x), comma));
    }
    s.push_str("}\n");

    // output dir is the primitive dir (parent of _gen); the runner sets CWD there or here.
    let target = Path::new("rust-output.json");
    fs::write(target, s).expect("write rust-output.json");
    println!("wrote rust-output.json (hand-port)");
}
